cl/lib/redis_utils.py

```diff
@@ -76,3 +76,24 @@ def delete_redis_semaphore(r: Union[str, Redis], key: str) -> None:
     if isinstance(r, str):
         r = make_redis_interface(r)
     r.delete(key)
+
+
+def check_or_create_email_sending_quota(r: Redis) -> None:
+    """Checks the value of the delivery_attempts key or creates it
+    if it's expired.
+
+    Args:
+        r (Redis): The Redis DB to connect to as a connection interface
+
+    Raises:
+        ValueError: if the counter is bigger than the threshold from the settings.
+    """
+
+    email_counter = r.get("email:delivery_attempts")
+    if email_counter:
+        if int(email_counter) >= settings.SENT_EMAILS_THRESHOLD:
+            raise ValueError("Emergency brake engaged")
+    else:
+        pipe = r.pipeline()
+        pipe.expire("email:delivery_attempts", 60 * 60 * 24)  # 24 hours period
+        pipe.execute()
```

Comment:
This is, I'm sorry to say, too simple. SES uses a rolling 24 hour period. From [the docs](https://docs.aws.amazon.com/ses/latest/dg/manage-sending-quotas.html):

> This quota is calculated on a rolling time period. Every time you try to send an email, Amazon SES determines the number of emails that you sent in the previous 24 hours. As long as the total number of emails that you have sent in the past 24 hours is less than this daily maximum, your send request is accepted and your email is sent.

Something tells me this is sort of a fun thing to figure out using redis.

Reply:
Oh yeah, I didn't think about it, but yes, it might be a problem. Once the key expires in Redis after 24 hours, the counter will start from scratch, while the 24-hour SES counter will be something totally different.

We could also use the `EmailSent` objects and get the count of objects created in the previous 24 hours for a more accurate count.

Or if we still want to use Redis a better approximation could be to hold a key for emails sent at each hour of the day. Then, add up all of them to get a more accurate count.

Reply:
I have to confess, I find this problem pretty fun to think about. I researched this a lot a year or two ago when I implemented the `@throttle_task` decorator for Celery tasks. 

This page has some really good explainers:

https://medium.com/@SaiRahulAkarapu/rate-limiting-algorithms-using-redis-eb4427b47e33

> If we still want to use Redis a better approximation could be to hold a key for emails sent at each hour of the day. Then, add up all of them to get a more accurate count.

Yeah, that could work with a pretty good balance between performance and accuracy, but I think the accuracy would be pretty low. For example, during this attack, we exceeded our limit in just a few hours. The accuracy could be improved by having per-minute keys instead, but that's a lot of keys.

> We could also use the EmailSent objects and get the count of objects created in the previous 24 hours for a more accurate count.

This is known as the sliding logs algorithm, and it could work, but it seems pretty inefficient to query postgresql in this way. I guess we could fix that by only querying it every ten emails or only when `unixtime % 10 == 0`, but that's a bummer.

The article above describes doing this with sorted sets in Redis (see the sliding logs section). I think that'd be pretty efficient and accurate, with the tradeoff of needing a somewhat large object in redis if we're ever under attack?

There are some other exotic algorithms for this that I came across. I'd be happy to look at others if y'all found one you liked better. I'm trying to ask myself WWAD (What Would Amazon Do?).

Reply:
It seems like this got stuck. Anything we need to clarify or do to un-stick it?

Reply:
What's the approach?

Reply:
We read the Medium article and saw that you were right about the sliding log algorithm, although it would be accurate, we would need to add a new item to the sorted list each time the system sends an email. If we use this algorithm as it is, we'll likely create a large set. For example, if we're ever under attack, the Redis set will have 50k elements (I also checked the SES dashboard and it showed we've sent 3400 emails in the last 24 hours).

In Alberto's approach, a temporary counter is used, and once it reaches a certain number of requests (for example, 10 requests), a new key can be created in the sorted set (as in the algorithm) and the counter can be restarted. Thus, the number of keys in the sorted set is reduced, and we can compute the real number of requests reliably (only multiplying x10 or by the upper limit we set for the temporary counter). The number of emails will be computed using the following formula:

``` python
email_count = temp_counter + key_count_in_set*upper_limit
```

Reply:
Interesting. I'll look forward to seeing the balance of approaches that you come up with.